presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java

```diff
@@ -0,0 +1,252 @@
+                        }
+                    }
+                    psplits.add(new PhoenixInputSplit(scans, regionSize, regionLocation));
+                }
+            }
+            return psplits;
+        }
+    }
+
+    private QueryPlan getQueryPlan(final JobContext context, final Configuration configuration)
+    {
+        Preconditions.checkNotNull(context);
+        try {
+            final String txnScnValue = configuration.get(PhoenixConfigurationUtil.TX_SCN_VALUE);
+            final String currentScnValue = configuration.get(PhoenixConfigurationUtil.CURRENT_SCN_VALUE);
+            final String tenantId = configuration.get(PhoenixConfigurationUtil.MAPREDUCE_TENANT_ID);
+            final Properties overridingProps = new Properties();
+            if (txnScnValue == null && currentScnValue != null) {
+                overridingProps.put(PhoenixRuntime.CURRENT_SCN_ATTRIB, currentScnValue);
+            }
+            if (tenantId != null && configuration.get(PhoenixRuntime.TENANT_ID_ATTRIB) == null) {
+                overridingProps.put(PhoenixRuntime.TENANT_ID_ATTRIB, tenantId);
+            }
+            try (final Connection connection = ConnectionUtil.getInputConnection(configuration, overridingProps);
+                    final Statement statement = connection.createStatement()) {
+                String selectStatement = PhoenixConfigurationUtil.getSelectStatement(configuration);
+                Preconditions.checkNotNull(selectStatement);
+
+                final PhoenixStatement pstmt = statement.unwrap(PhoenixStatement.class);
+                // Optimize the query plan so that we potentially use secondary indexes
+                final QueryPlan queryPlan = pstmt.optimizeQuery(selectStatement);
+                final Scan scan = queryPlan.getContext().getScan();
+
+                // since we can't set a scn on connections with txn set TX_SCN attribute so that the max time range is set by BaseScannerRegionObserver
+                if (txnScnValue != null) {
+                    scan.setAttribute(BaseScannerRegionObserver.TX_SCN, Bytes.toBytes(Long.valueOf(txnScnValue)));
+                }
+
+                // setting the snapshot configuration
+                String snapshotName = configuration.get(PhoenixConfigurationUtil.SNAPSHOT_NAME_KEY);
+                if (snapshotName != null) {
+                    PhoenixConfigurationUtil.setSnapshotNameKey(
+                            queryPlan.getContext().getConnection().getQueryServices().getConfiguration(),
+                            snapshotName);
+                }
+
+                // Initialize the query plan so it sets up the parallel scans
+                queryPlan.iterator(MapReduceParallelScanGrouper.getInstance());
+                return queryPlan;
+            }
+        }
+        catch (Exception e) {
+            log.error(String.format("Failed to get the query plan with error [%s]", e.getMessage()));
```

Comment:
log-and-throw is an anti-pattern that we avoid in Presto. It results in the same error getting logged in multiple places, which is confusing. Either log or throw, but not both. In this case, it's probably best to re-throw like this:
```java
throw new PrestoException(PHOENIX_ERROR, "Failed to get the Phoenix query plan", e);

Reply:
Also, create separate error codes when possible. For example, `PHOENIX_QUERY_PLAN_ERROR` might be appropriate here. It helps administrators to understand why understand why queries failed, count query failures by error type, etc.